include/envoy/ssl/handshaker.h

```diff
@@ -0,0 +1,101 @@
+#pragma once
+
+#include "envoy/api/api.h"
+#include "envoy/common/pure.h"
+#include "envoy/config/typed_config.h"
+#include "envoy/network/transport_socket.h"
+#include "envoy/ssl/socket_state.h"
+
+#include "openssl/ssl.h"
+
+namespace Envoy {
+namespace Ssl {
+
+class HandshakerCallbacks {
+public:
+  virtual ~HandshakerCallbacks() = default;
+
+  /**
+   * Hands off the internally-held SSL object for external manipulation.
+   */
+  virtual bssl::UniquePtr<SSL> HandOff() PURE;
+  /**
+   * Accepts an SSL object for internal storage.
+   */
+  virtual void HandBack(bssl::UniquePtr<SSL> ssl) PURE;
+  /**
+   * Called when a handshake is successfully performed.
+   */
+  virtual void OnSuccessCb(SSL* ssl) PURE;
+  /**
+   * Called when a handshake fails.
+   */
+  virtual void OnFailureCb() PURE;
+};
+
+class Handshaker {
+public:
+  virtual ~Handshaker() = default;
+
+  // The initialize method must be called before
+  // doHandshake() is called.
+  virtual void initialize(SSL& ssl) PURE;
+
+  /**
+   * Do the handshake.
+   *
+   *  * |state| is a mutable reference.
+   *  * |ssl| is a pointer, and this method may be called with |ssl| == nullptr,
+   *    for example during a reentrant call to doHandshake() during a period
+   *    when the handshaker has handed off the SSL*.
+   *  * |callbacks| may not exist throughout the lifetime of the Handshaker, and
+   *    should not be stored in an implementation.
+   */
+  virtual Network::PostIoAction doHandshake(SocketState& state, SSL* ssl,
+                                            HandshakerCallbacks& callbacks) PURE;
+
+  /**
+   * Set an internal pointer to the Network::TransportSocketCallbacks struct.
+   * Depending on impl, these callbacks can be invoked to access connection
+   * state, raise connection events, etc.
+   */
+  virtual void setTransportSocketCallbacks(Network::TransportSocketCallbacks& callbacks) PURE;
+};
+
+using HandshakerPtr = std::unique_ptr<Handshaker>;
+
+class HandshakerFactoryContext {
+public:
+  virtual ~HandshakerFactoryContext() = default;
+
+  /**
+   * @return reference to the Api object
+   */
+  virtual Api::Api& api() PURE;
+
+  /**
+   * The list of supported protocols exposed via ALPN, from ContextConfig.
+   */
+  virtual absl::string_view alpnProtocols() const PURE;
+};
+
+class HandshakerFactory : public Config::TypedFactory {
+public:
+  virtual HandshakerPtr createHandshaker(HandshakerFactoryContext& context) PURE;
+
+  std::string category() const override { return "envoy.tls_handshakers"; }
+
+  /**
+   * Set a config for use when creating handshakers.
+   */
+  virtual void setConfig(ProtobufTypes::MessagePtr message) PURE;
```

Comment:
Please take a look at the way factories, config, etc and follow that pattern. There's basically two levels of factories: the singleton (for each filter/plugin/type) that goes in the registry, which has a single method like "create config factory from config", which produces something like this `HandshakerFactory`. But then the `HandshakerFactory` is constructed with it's config, so there's no need for a `setConfig()` method. See for example `include/envoy/server/filter_config.h`.

Reply:
I definitely want to follow the Envoy factory patterns more idiomatically, but I can't figure out what you mean here -- could you give an explicit link into `filter_config.h` for the pattern to follow? I don't see where I can configure a Factory to accept a config at ctor time, since the Factory structs (as far as I can tell) get implicitly created way before the config is available.

Reply:
I think I've achieved a HandshakerFactoryCb model which has the desired goal (one-time early config validation). There is some weirdness here -- if I had made the HandshakerFactor impls templated per-config type, then I would not be able to use getAndCheckFactory. But I'm now much closer to following the pattern that `filter_config.h` establishes for early config validation, PTAL.